OpenSSL no_deprecated

**Describe the bug**
Poco fails to build against OpenSSL 3.6.3 that was built with no-deprecated=true

**To Reproduce**
Will investigate small sample project to demonstrate build missing.

openssl/3.6.3
poco/1.15.2

Conan config 

[options]
openssl/*:shared=False
openssl/*:no_deprecated=True
openssl/*:no_legacy=True
poco/*:shared=False

**Expected behavior**
expand current identification in usage of OpenSSL items marked deprecated so that no-deprecated may be used.

**Logs**

/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SSLManager.cpp: In static member function ‘static int Poco::Net::SSLManager::verifyOCSPResponseCallback(SSL*, void*)’:
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SSLManager.cpp:305:20: error: ‘SSL_get_peer_certificate’ was not declared in this scope
  X509* pPeerCert = SSL_get_peer_certificate(pSSL);
                    ^~~~~~~~~~~~~~~~~~~~~~~~
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SSLManager.cpp:305:20: note: suggested alternative: ‘SSL_get1_peer_certificate’
  X509* pPeerCert = SSL_get_peer_certificate(pSSL);
                    ^~~~~~~~~~~~~~~~~~~~~~~~
                    SSL_get1_peer_certificate
NetSSL_OpenSSL/CMakeFiles/NetSSL.dir/build.make:327: recipe for target 'NetSSL_OpenSSL/CMakeFiles/NetSSL.dir/src/SSLManager.cpp.o' failed
gmake[2]: *** [NetSSL_OpenSSL/CMakeFiles/NetSSL.dir/src/SSLManager.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp: In member function ‘long int Poco::Net::SecureSocketImpl::verifyPeerCertificateImpl(const string&)’:
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:490:20: error: ‘::SSL_get_peer_certificate’ has not been declared
  ::X509* pCert = ::SSL_get_peer_certificate(_pSSL);
                    ^~~~~~~~~~~~~~~~~~~~~~~~
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:490:20: note: suggested alternative: ‘SSL_get1_peer_certificate’
  ::X509* pCert = ::SSL_get_peer_certificate(_pSSL);
                    ^~~~~~~~~~~~~~~~~~~~~~~~
                    SSL_get1_peer_certificate
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp: In member function ‘X509* Poco::Net::SecureSocketImpl::peerCertificate() const’:
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:519:12: error: ‘::SSL_get_peer_certificate’ has not been declared
   return ::SSL_get_peer_certificate(_pSSL);
            ^~~~~~~~~~~~~~~~~~~~~~~~
/builds/.conan2/p/b/poco6758757fc253f/b/src/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:519:12: note: suggested alternative: ‘SSL_get1_peer_certificate’
   return ::SSL_get_peer_certificate(_pSSL);
            ^~~~~~~~~~~~~~~~~~~~~~~~
            SSL_get1_peer_certificate


**Please add relevant environment information:**
 - OS Type and Version: Linux, SUSE 12 SP5
 - POCO Version: 1.15.2
 - Built from conan2 recipe,  gcc8

https://github.com/pocoproject/poco/blob/1dd27806a26477c0771a4f863039af4c18696afe/NetSSL_OpenSSL/src/SSLManager.cpp#L304
https://github.com/pocoproject/poco/blob/1dd27806a26477c0771a4f863039af4c18696afe/NetSSL_OpenSSL/src/SSLManager.cpp#L15-L17
https://github.com/pocoproject/poco/blob/1dd27806a26477c0771a4f863039af4c18696afe/NetSSL_OpenSSL/src/SecureSocketImpl.cpp#L489
https://github.com/pocoproject/poco/blob/1dd27806a26477c0771a4f863039af4c18696afe/NetSSL_OpenSSL/src/SecureSocketImpl.cpp#L518

**Additional context**
AI Overview                 
In OpenSSL 3.0, SSL_get_peer_certificate has been deprecated. It is replaced by SSL_get1_peer_certificate or SSL_get0_peer_certificate, which handle certificate reference counting explicitly.If you are upgrading an older codebase to OpenSSL 3, you will likely see compilation warnings or linker errors like undefined reference to SSL_get_peer_certificate if your environment compiles with OPENSSL_NO_DEPRECATED enabled.

Recommended Replacements
You should modify your code to use one of the two new functions depending on how you manage memory:
SSL_get1_peer_certificate() [1](https://manpages.debian.org/experimental/libssl-doc/SSL_get_peer_certificate.3ssl.en.html)

Behavior: Identical to the deprecated function.
Memory Management: It increments the reference count of the X509 object.
Action Required: You must explicitly free the pointer using X509_free() when you are done with it to avoid memory leaks.

SSL_get0_peer_certificate()
Behavior: Returns a pointer to the internal certificate object.
Memory Management: It does not increment the reference count.
Action Required: Do not free this pointer. The memory is tied to the lifecycle of the SSL connection object.

Working on updated Conan recipe for 1.15.3, compile on Windows vc142/VS2019  - issue is not platform specific.

>  VerificationErrorArgs.cpp
C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\src\NetSSL_OpenSSL\src\SSLManager.cpp(305,20): error C3861: 'SSL_get_peer_certificate': identifier not found [C:\U
sers\gdomjan\.conan2\p\b\pococ3274fbf8ed9e\b\build\NetSSL_OpenSSL\NetSSL.vcxproj]
  X509Certificate.cpp
C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\src\NetSSL_OpenSSL\src\SecureSocketImpl.cpp(490,20): error C2039: 'SSL_get_peer_certificate': is not a member of '
`global namespace'' [C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\build\NetSSL_OpenSSL\NetSSL.vcxproj]
C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\src\NetSSL_OpenSSL\src\SecureSocketImpl.cpp(490,44): error C3861: 'SSL_get_peer_certificate': identifier not found
 [C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\build\NetSSL_OpenSSL\NetSSL.vcxproj]
C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\src\NetSSL_OpenSSL\src\SecureSocketImpl.cpp(519,12): error C2039: 'SSL_get_peer_certificate': is not a member of '
`global namespace'' [C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\build\NetSSL_OpenSSL\NetSSL.vcxproj]
C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\src\NetSSL_OpenSSL\src\SecureSocketImpl.cpp(519,36): error C3861: 'SSL_get_peer_certificate': identifier not found
 [C:\build\.conan2\p\b\pococ3274fbf8ed9e\b\build\NetSSL_OpenSSL\NetSSL.vcxproj]

Patching via preprocessor allows workaround

``` python
        if "openssl" in self.dependencies and Version(self.dependencies["openssl"].ref.version) >= "3":
            tc.preprocessor_definitions["SSL_get_peer_certificate"] = "SSL_get1_peer_certificate"
```

--- a/.github/path-filters.yml
+++ b/.github/path-filters.yml
@@ -34,3 +34,8 @@ data:
 
 sqlparser:
   - 'Data/SQLParser/**'
+
+crypto_netssl:
+  - 'Crypto/**'
+  - 'NetSSL_OpenSSL/**'
+  - 'JWT/**'

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -100,6 +100,7 @@ jobs:
       arm_cross: ${{ steps.filter.outputs.arm_cross }}
       data: ${{ steps.filter.outputs.data }}
       sqlparser: ${{ steps.filter.outputs.sqlparser }}
+      crypto_netssl: ${{ steps.filter.outputs.crypto_netssl }}
     steps:
       - uses: actions/checkout@v5
       - uses: dorny/paths-filter@v4
@@ -232,6 +233,36 @@ jobs:
             cd cmake-build &&
             PWD=`pwd`
             ctest ${{ env.POCO_CTEST_COMMON }} --parallel $(nproc) -R "^(Foundation|Net|Crypto|Util|XML|JSON|Data|DataSQLite)$"
+      - uses: ./.github/actions/upload-test-report
+
+  # OpenSSL no-deprecated probe: build Crypto/NetSSL/JWT with deprecated
+  # OpenSSL declarations hidden (#5415). Debug + -D_DEBUG so that
+  # poco_assert_dbg expressions compile too.
+  linux-gcc-cmake-openssl-no-deprecated:
+    runs-on: ubuntu-24.04
+    needs: changes
+    if: |
+      needs.changes.outputs.crypto_netssl == 'true' ||
+      needs.changes.outputs.ci_core == 'true' ||
+      github.event_name == 'push' || github.event_name == 'workflow_dispatch'
+    steps:
+      - uses: actions/checkout@v5
+      - run: sudo apt -y update && sudo apt -y install cmake ninja-build libssl-dev
+      - run: >-
+          cmake -S. -Bcmake-build -GNinja -DCMAKE_BUILD_TYPE=Debug
+          -DCMAKE_CXX_FLAGS="-DOPENSSL_NO_DEPRECATED -D_DEBUG"
+          -DPOCO_MINIMAL_BUILD=ON -DENABLE_TESTS=ON
+          -DENABLE_CRYPTO=ON -DENABLE_NETSSL=ON -DENABLE_JWT=ON
+      - run: cmake --build cmake-build --target all --parallel $(nproc)
+      - uses: ./.github/actions/retry-action
+        with:
+          timeout_minutes: 30
+          max_attempts: 3
+          retry_on: any
+          command: >-
+            cd cmake-build &&
+            PWD=`pwd`
+            ctest ${{ env.POCO_CTEST_COMMON }} --parallel $(nproc) -R "^(Crypto|JWT)$"
       - uses: ./.github/actions/upload-test-report
 
   # POCO_SOO=OFF probe: Foundation only.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -235,9 +235,9 @@ jobs:
             ctest ${{ env.POCO_CTEST_COMMON }} --parallel $(nproc) -R "^(Foundation|Net|Crypto|Util|XML|JSON|Data|DataSQLite)$"
       - uses: ./.github/actions/upload-test-report
 
-  # OpenSSL no-deprecated probe: build Crypto/NetSSL/JWT with deprecated
-  # OpenSSL declarations hidden (#5415). Debug + -D_DEBUG so that
-  # poco_assert_dbg expressions compile too.
+  # OpenSSL no-deprecated probe: build and test Crypto/NetSSL/JWT with
+  # deprecated OpenSSL declarations hidden. Debug + -D_DEBUG so that
+  # poco_assert_dbg expressions are compiled as well.
   linux-gcc-cmake-openssl-no-deprecated:
     runs-on: ubuntu-24.04
     needs: changes
@@ -262,7 +262,7 @@ jobs:
           command: >-
             cd cmake-build &&
             PWD=`pwd`
-            ctest ${{ env.POCO_CTEST_COMMON }} --parallel $(nproc) -R "^(Crypto|JWT)$"
+            ctest ${{ env.POCO_CTEST_COMMON }} --parallel $(nproc) -R "^(Crypto|NetSSL|JWT)$"
       - uses: ./.github/actions/upload-test-report
 
   # POCO_SOO=OFF probe: Foundation only.

--- a/Crypto/include/Poco/Crypto/EVPPKey.h
+++ b/Crypto/include/Poco/Crypto/EVPPKey.h
@@ -202,8 +202,13 @@ private:
 		const std::string& keyFile,
 		const std::string& pass = "")
 	{
+#ifndef OPENSSL_NO_DEPRECATED_3_0
 		poco_assert_dbg (((typeid(K*) == typeid(RSA*) || typeid(K*) == typeid(EC_KEY*)) && getFunc) ||
 						((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc));
+#else
+		// RSA/EC_KEY types are not declared; only EVP_PKEY loading is possible
+		poco_assert_dbg ((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc);
+#endif
 		poco_check_ptr (ppKey);
 		poco_assert_dbg (!*ppKey);
 
@@ -274,8 +279,13 @@ private:
 		std::istream* pIstr,
 		const std::string& pass = "")
 	{
+#ifndef OPENSSL_NO_DEPRECATED_3_0
 		poco_assert_dbg (((typeid(K*) == typeid(RSA*) || typeid(K*) == typeid(EC_KEY*)) && getFunc) ||
 						((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc));
+#else
+		// RSA/EC_KEY types are not declared; only EVP_PKEY loading is possible
+		poco_assert_dbg ((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc);
+#endif
 		poco_check_ptr(ppKey);
 		poco_assert_dbg(!*ppKey);
 

--- a/Crypto/include/Poco/Crypto/EVPPKey.h
+++ b/Crypto/include/Poco/Crypto/EVPPKey.h
@@ -29,6 +29,7 @@
 #include <openssl/pem.h>
 #include <sstream>
 #include <map>
+#include <type_traits>
 
 
 namespace Poco::Crypto {
@@ -202,12 +203,13 @@ private:
 		const std::string& keyFile,
 		const std::string& pass = "")
 	{
-#ifndef OPENSSL_NO_DEPRECATED_3_0
-		poco_assert_dbg (((typeid(K*) == typeid(RSA*) || typeid(K*) == typeid(EC_KEY*)) && getFunc) ||
-						((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc));
+#if POCO_OPENSSL_VERSION_PREREQ(3, 0, 0)
+		// Native key extraction is only compiled for OpenSSL < 3.0; the RSA and
+		// EC_KEY types are not even declared when deprecated APIs are disabled.
+		poco_assert_dbg ((std::is_same_v<K, EVP_PKEY> && !getFunc));
 #else
-		// RSA/EC_KEY types are not declared; only EVP_PKEY loading is possible
-		poco_assert_dbg ((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc);
+		poco_assert_dbg (((std::is_same_v<K, RSA> || std::is_same_v<K, EC_KEY>) && getFunc) ||
+						(std::is_same_v<K, EVP_PKEY> && !getFunc));
 #endif
 		poco_check_ptr (ppKey);
 		poco_assert_dbg (!*ppKey);
@@ -242,7 +244,7 @@ private:
 						}
 						else
 						{
-							poco_assert_dbg (typeid(K*) == typeid(EVP_PKEY*));
+							poco_assert_dbg ((std::is_same_v<K, EVP_PKEY>));
 							*ppKey = (K*)pKey;
 						}
 						if (!*ppKey) goto error;
@@ -279,12 +281,13 @@ private:
 		std::istream* pIstr,
 		const std::string& pass = "")
 	{
-#ifndef OPENSSL_NO_DEPRECATED_3_0
-		poco_assert_dbg (((typeid(K*) == typeid(RSA*) || typeid(K*) == typeid(EC_KEY*)) && getFunc) ||
-						((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc));
+#if POCO_OPENSSL_VERSION_PREREQ(3, 0, 0)
+		// Native key extraction is only compiled for OpenSSL < 3.0; the RSA and
+		// EC_KEY types are not even declared when deprecated APIs are disabled.
+		poco_assert_dbg ((std::is_same_v<K, EVP_PKEY> && !getFunc));
 #else
-		// RSA/EC_KEY types are not declared; only EVP_PKEY loading is possible
-		poco_assert_dbg ((typeid(K*) == typeid(EVP_PKEY*)) && !getFunc);
+		poco_assert_dbg (((std::is_same_v<K, RSA> || std::is_same_v<K, EC_KEY>) && getFunc) ||
+						(std::is_same_v<K, EVP_PKEY> && !getFunc));
 #endif
 		poco_check_ptr(ppKey);
 		poco_assert_dbg(!*ppKey);
@@ -314,7 +317,7 @@ private:
 						}
 						else
 						{
-							poco_assert_dbg (typeid(K*) == typeid(EVP_PKEY*));
+							poco_assert_dbg ((std::is_same_v<K, EVP_PKEY>));
 							*ppKey = (K*)pKey;
 						}
 						if (!*ppKey) goto error;

--- a/NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h
+++ b/NetSSL_OpenSSL/include/Poco/Net/SecureSocketImpl.h
@@ -190,7 +190,9 @@ public:
 		/// Returns the underlying socket descriptor.
 
 	X509* peerCertificate() const;
-		/// Returns the peer's certificate.
+		/// Returns the peer's certificate, or a null pointer if
+		/// no certificate is available. The caller is responsible
+		/// for releasing the returned reference with X509_free().
 
 	Context::Ptr context() const;
 		/// Returns the SSL context used for this socket.

--- a/NetSSL_OpenSSL/src/SSLManager.cpp
+++ b/NetSSL_OpenSSL/src/SSLManager.cpp
@@ -301,7 +301,11 @@ int SSLManager::verifyOCSPResponseCallback(SSL* pSSL, void* arg)
 		return 0;
 	}
 
+#if POCO_OPENSSL_VERSION_PREREQ(3, 0, 0)
+	X509* pPeerCert = SSL_get1_peer_certificate(pSSL);
+#else
 	X509* pPeerCert = SSL_get_peer_certificate(pSSL);
+#endif
 	if (!pPeerCert)
 	{
 		OCSP_BASICRESP_free(pBasicResp);

--- a/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
+++ b/NetSSL_OpenSSL/src/SecureSocketImpl.cpp
@@ -486,7 +486,11 @@ long SecureSocketImpl::verifyPeerCertificateImpl(const std::string& hostName)
 		return X509_V_OK;
 	}
 
+#if POCO_OPENSSL_VERSION_PREREQ(3, 0, 0)
+	::X509* pCert = ::SSL_get1_peer_certificate(_pSSL);
+#else
 	::X509* pCert = ::SSL_get_peer_certificate(_pSSL);
+#endif
 	if (pCert)
 	{
 		X509Certificate cert(pCert);
@@ -514,10 +518,14 @@ X509* SecureSocketImpl::peerCertificate() const
 {
 	LockT l(_mutex);
 
-	if (_pSSL)
-		return ::SSL_get_peer_certificate(_pSSL);
-	else
+	if (_pSSL == nullptr)
 		return nullptr;
+
+#if POCO_OPENSSL_VERSION_PREREQ(3, 0, 0)
+	return ::SSL_get1_peer_certificate(_pSSL);
+#else
+	return ::SSL_get_peer_certificate(_pSSL);
+#endif
 }
 
 

